File: coding_problems/dp/knapsack/subset_sum.py
def __subset__(numbers, size,target, mem):
    if(size == 0):
        return false
    if(target == 0):
        return true;
    if(mem[size][target]!= -1): 
        return mem[size][target]
    last_num = numbers[size]
    if(last_num>target):
        mem[size][target] = __subset__(numbers.pop(size),size-1,target)
    else:
        mem[size][target] = __subset__(numbers.pop(size),size -1, target + last_num) or __subset__(numbers.pop(size),size-1,target)


def subset( numbers,  target):
    mem[numbers+1][target+1]
    for i in range(0,numbers+1):
        for j in range(0,target+1):
            mem[i][j] = -1
    size = len(numbers)
    return __subset__(numbers,size,target,mem)

numbers = [2,3,7,8,10]
target = 11

print(subset(numbers,target))

File: coding_problems/dp/knapsack/subset_sum.cpp
#include<bits/stdc++.h>
using namespace std;

int __subset__(vector<int> numbers,int size,int target, int memo[6][12])
{
    if(size == 0)
        return 0;
    if(target == 0)
        return 1;
    int x =numbers[size - 1];

    numbers.pop_back();
    if(memo[size][target]!= -1) return memo[size][target];
    if(x>target)
    {
    memo[size][target] = __subset__(numbers,size -1,target,memo);
    }
    else
    {
        memo[size][target] = (__subset__(numbers,size -1,target -x,memo) || __subset__(numbers,size -1,target,memo));
    }
    return memo[size][target];
}

bool subset(vector<int> numbers,int target)
{
    
    int size = numbers.size();
    int memo[6][12] = {-1};
    if(__subset__(numbers,6,12,memo)==1)
        return true;
    else 
        return false;
}

int main()
{
    vector<int> numbers {2,3,7,8,10};
    int target = 11;
    cout<<true;
    cout<<subset(numbers,target);
	return 0;
}